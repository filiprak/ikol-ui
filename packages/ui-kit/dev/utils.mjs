import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
