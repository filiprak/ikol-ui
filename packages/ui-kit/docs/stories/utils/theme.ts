import { useTheme } from '@ui/composables';
import { createIkolUI } from '@ui/index';
import { App } from 'vue';
import { StoryContext, VueRenderer } from '@storybook/vue3';

export function setupTheme(context: any) {
    const theme = useTheme();

    theme.is_dark.value = context?.globals?.theme === 'dark';

    return theme;
}

export function setupApp(app: App, context?: StoryContext<VueRenderer>) {
    app.use(createIkolUI());
}
