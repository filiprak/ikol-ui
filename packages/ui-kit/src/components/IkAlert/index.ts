import IkAlert from './IkAlert.vue';
export interface IkAlert extends InstanceType<typeof IkAlert> { };
export { IkAlert };
