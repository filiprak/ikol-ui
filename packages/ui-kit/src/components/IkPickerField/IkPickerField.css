.ik-picker-field-item--default,
.ik-picker-field-item--empty {
    background-color: var(--surface-darken-color);
    color: var(--primary-color);
}

.ik-picker-field-item--empty .ik-list-item__content {
    color: var(--on-surface-darken-color-3);
}

.ik-picker-field-item--default .ik-list-item__content {
    color: var(--on-surface-darken-color-1);
}

.ik-picker-field-item--disabled {
    opacity: 0.65;
    pointer-events: none;
}

.ik-picker-field-item {
    border-radius: 8px;
    padding: 10px;
    cursor: pointer;
}

.ik-picker-field__item {
    cursor: pointer;
}

.ik-picker-field__item:hover {
    background-color: var(--primary-color-o1);
    color: var(--primary-color);
}