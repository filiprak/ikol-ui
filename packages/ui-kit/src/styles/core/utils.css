/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
