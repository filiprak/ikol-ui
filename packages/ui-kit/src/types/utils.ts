import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
