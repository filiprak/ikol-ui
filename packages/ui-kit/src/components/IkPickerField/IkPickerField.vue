<template>
    <IkFormField class="ik-picker-field"
                 :errors="filtered_errors"
                 v-bind="$props">
        <IkSelect ref="select"
                  :model-value="props.modelValue"
                  :value_key="props.value_key"
                  :text_key="props.text_key"
                  :text_expression="props.text_expression"
                  :items="props.items"
                  variant="filled"
                  :disabled="props.disabled"
                  @update:modelValue="emitVModel">
            <template #activator="{ on, selected }">
                <slot name="activator"
                      :on="on"
                      :selected="selected">
                    <IkPickerFieldItem :text_expression="props.text_expression"
                                       :model-value="selected"
                                       :disabled="props.disabled"
                                       :text_key="props.text_key"
                                       v-on="on">
                        <slot name="action-text"></slot>
                    </IkPickerFieldItem>
                </slot>
            </template>
            <template #item="item">
                <IkListItem no_wrap
                            grow_content
                            class="ik-picker-field__item">
                    <template v-if="$slots.icon"
                              #prepend>
                        <slot name="icon"
                              :item="item.item" />
                    </template>
                    <slot name="text">
                        <span>{{ getItemText(item.item) }}</span>
                    </slot>
                </IkListItem>
            </template>
        </IkSelect>
    </IkFormField>
</template>

<script setup lang="ts" generic="ItemT extends {}, ValueT = ItemT">

import './IkPickerField.css';

import { IkFormField } from '../IkForm';
import { IkListItem } from '../IkList';
import { IkSelect, type IkSelectItemLoader } from '../IkSelect';
import IkPickerFieldItem from './IkPickerFieldItem.vue';
import { type IkFieldProps, getFieldPropDefaults, useFormField } from '@ui/composables/form';
import { ref } from 'vue';
import type { IkKeysByType } from '@ui/types/utils';

const props = withDefaults(
    defineProps<IkFieldProps & {
        modelValue?: ValueT;
        items?: ItemT[] | IkSelectItemLoader<ItemT>;
        value_key?: IkKeysByType<ItemT, ValueT>;
        text_key?: keyof ItemT,
        text_expression?: (item: ItemT) => string;
        placeholder?: string;
        disabled?: boolean;
    }>(),
    {
        ...getFieldPropDefaults(),
        items: () => [],
        value_key() {
            return 'value' as unknown as IkKeysByType<ItemT, ValueT>;
        },
        text_key() {
            return 'text' as unknown as keyof ItemT;
        },
        placeholder: '[[_*en*Not set_]]',
        disabled: false,
    }
);

defineOptions({
    inheritAttrs: false,
});

const field = ref<IkFormField>();
const select = ref<IkSelect>();

const {
    filtered_errors,
    computed_variant,
    computed_skeleton,
    computed_view_mode,
    emitVModel,
    ...methods

} = useFormField({
    field,
    props,
});

defineExpose({
    ...methods,
    focus,
});

const getItemText = (data: ItemT) => {
    if (props.text_expression) {
        return props.text_expression(data);
    } else {
        return data[props.text_key as keyof ItemT] || '';
    }
};

</script>
