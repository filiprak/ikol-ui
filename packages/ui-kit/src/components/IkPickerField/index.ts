import type { IkExtractExposed } from '@ui/types/utils';
import IkPickerField from './IkPickerField.vue';
import IkPickerFieldItem from './IkPickerFieldItem.vue';
export interface IkPickerField extends IkExtractExposed<typeof IkPickerField> { };
export interface IkPickerFieldItem extends IkExtractExposed<typeof IkPickerFieldItem> { };
export { IkPickerField, IkPickerFieldItem };
