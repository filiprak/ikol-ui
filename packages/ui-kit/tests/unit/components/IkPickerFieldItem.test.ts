import { IkPickerFieldItem } from '@ui/components/IkPickerField';
import { type RenderOptions, renderDeep } from '@ui/tests/tools/utils';

describe('IkPickerFieldItem', () => {
    const getInstance = (options: RenderOptions<typeof IkPickerFieldItem> = {}) => renderDeep(IkPickerFieldItem, options);

    it('renders with default state', () => {
        const wrapper = getInstance({
            props: {},
        });

        expect(wrapper.text()).toContain('Not set');
    });

    it('renders with value', () => {
        const wrapper = getInstance({
            props: {
                modelValue: {
                    code: 'COMPLETE',
                    color: 'BBC7D0',
                    text: 'Complete',
                },
            },
        });

        expect(wrapper.text()).toContain('Complete');
    });

    it('renders with text expression', () => {
        const wrapper = getInstance({
            props: {
                modelValue: {
                    code: 'COMPLETE',
                    color: 'BBC7D0',
                    text: 'Complete',
                },
                text_expression: (value: any) => {
                    return value.text.toUpperCase();
                },
            },
        });

        expect(wrapper.text()).toContain('COMPLETE');
    });

    it('renders disabled', () => {
        const wrapper = getInstance({
            props: {
                modelValue: {
                    code: 'COMPLETE',
                    color: 'BBC7D0',
                    text: 'Complete',
                },
                disabled: true,
                text_expression: (value: any) => {
                    return value.text.toUpperCase();
                },
            },
        });

        expect(wrapper.html()).toContain('ik-picker-field-item--disabled');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders action text as CHOOSE when modelValue is falsy', () => {
        const wrapper = getInstance({
            props: {},
        });

        expect(wrapper.text()).toContain('CHOOSE');
    });

    it('renders action text as UPDATE when modelValue is truthy', () => {
        const wrapper = getInstance({
            props: {
                modelValue: {
                    code: 'COMPLETE',
                    color: 'BBC7D0',
                    text: 'Complete',
                },
            },
        });

        expect(wrapper.text()).toContain('UPDATE');
    });

    it('renders custom action text through slot', () => {
        const wrapper = getInstance({
            props: {
                modelValue: {
                    code: 'COMPLETE',
                    color: 'BBC7D0',
                    text: 'Complete',
                },
            },
            slots: {
                'action-text': '<span>Custom Action</span>',
            },
        });

        expect(wrapper.text()).toContain('Custom Action');
    });
});
