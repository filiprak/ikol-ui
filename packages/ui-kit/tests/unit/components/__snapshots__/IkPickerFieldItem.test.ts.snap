// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkPickerFieldItem renders disabled 1`] = `
<div
  class="ik-list-item ik-list-item--disabled ik-picker-field-item ik-picker-field-item--disabled"
  grow_content=""
  no_wrap=""
  style="background-color: rgb(226, 230, 232); color: rgb(141, 156, 167);"
>
  <div
    class="ik-list-item__prepend"
  >
    
    
    <div
      class="ik-color-swatch ik-color-swatch--round ik-color-swatch--md"
      style="background-color: rgb(187, 199, 208);"
    />
    
    
  </div>
  <div
    class="ik-list-item__content"
  >
    <div
      class="ik-list-item__title"
    >
      
      
      COMPLETE
      
      
    </div>
    <!--v-if-->
  </div>
  <div
    class="ik-list-item__append"
  >
    
    
    [[_*en*UPDATE_]]
    
    
  </div>
</div>
`;
